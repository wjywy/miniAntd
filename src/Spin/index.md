---
title: Spin 加载中
nav:
  title: '组件'
  path: /components
---
# Spin 加载中

用于页面和区块的加载中状态。

## 基本类型

<code src="./demos/basic.tsx">基本用法</code>
<code src="./demos/size.tsx">各种大小</code>

