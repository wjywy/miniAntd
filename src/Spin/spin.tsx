import classNames from 'classnames';
import React from 'react';

const SpinSizes = ['small', 'default', 'large'] as const;
export type SpinSize = typeof SpinSizes[number];
export type SpinIndicator = React.ReactElement<HTMLElement>;

export interface SpinProps {
  prefixCls?: string;
  className?: string;
  spinning?: boolean;
  style?: React.CSSProperties;
  size?: SpinSize;
  tip?: React.ReactNode;
  delay?: number;
  wrapperClassName?: string;
  indicator?: SpinIndicator;
  children?: React.ReactNode;
}

export interface SpinClassProps extends SpinProps {
  hashId: string;
  spinPrefixCls: string;
}

export type SpinFCType = React.FC<SpinProps> & {
  setDefaultIndicator: (indicator: React.ReactNode) => void;
};


function renderIndicator(prefixCls: string, props: SpinProps): React.ReactNode {
  const { indicator } = props;
  const dotClassName = `${prefixCls}-dot`;

  if (indicator === null) {
    return null;
  }

  return (
    <span className={classNames(dotClassName, `${prefixCls}-dot-spin`)}>
      <i className={`${prefixCls}-dot-item`} />
      <i className={`${prefixCls}-dot-item`} />
      <i className={`${prefixCls}-dot-item`} />
      <i className={`${prefixCls}-dot-item`} />
    </span>
  );
}

function shouldDelay(spinning?: boolean, delay?: number): boolean {
  return !!spinning && !!delay && !isNaN(Number(delay));
}

const Spin: React.FC<SpinProps> = (props) => {
  const {
    prefixCls = 'spin',
    spinning: customSpinning = true,
    delay = 0,
    className,
    size = 'default',
    tip,
    style,
    ...restProps
  } = props;

  const [spinning] = React.useState<boolean>(
    () => customSpinning && !shouldDelay(customSpinning, delay),
  );

  const spinClassName = classNames(
    prefixCls,
    {
      [`${prefixCls}-sm`]: size === 'small',
      [`${prefixCls}-lg`]: size === 'large',
      [`${prefixCls}-spinning`]: spinning,
      [`${prefixCls}-show-text`]: !!tip,
      // [`${prefixCls}-rtl`]: direction === 'rtl',
    },
    className,
  );

  return (
    <div
      {...restProps}
      style={style}
      className={spinClassName}
      aria-live="polite"
      aria-busy={spinning}
    >
      {renderIndicator(prefixCls, props)}
      {tip ? <div className={`${prefixCls}-text`}>{tip}</div> : null}
    </div>
  );
};

export default Spin;