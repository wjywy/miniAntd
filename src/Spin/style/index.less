.spin{
  margin-top: 10px;
  
.spin-dot {
  display: block;
  position: relative;
  top: calc(50% - 18px);
  margin: 0 auto;
  width: 36px;
  height: 36px;

  /* Safari 和 Chrome */
  transform: rotate(45deg);

  /* Opera */
  animation: rotate 1.2s linear infinite;

  @keyframes rotate {
    100% {
      transform: rotate(405deg);
    }
  }

  .spin-dot-item {
    // 单个圆点样式
    position: absolute;
    width: 10px;
    height: 10px;
    background: lightskyblue;
    border-radius: 50%;
    opacity: .3;
    animation: spin-move 1s linear infinite alternate;

    @keyframes spin-move {
      100% {
        opacity: 1;
      }
    }
  }

  .uspin-dot-item:first-child {
    top: 0;
    left: 0;
  }

  .spin-dot-item:nth-child(2) {
    top: 0;
    right: 0;
    animation-delay: .4s;
  }

  .spin-dot-item:nth-child(3) {
    bottom: 0;
    right: 0;
    animation-delay: .8s;
  }

  .spin-dot-item:last-child {
    bottom: 0;
    left: 0;
    animation-delay: 1.2s;
  }
}
}

.spin-sm .spin-dot{
  width: 20px;
  height: 20px;

  .spin-dot-item {
    width: 7px;
    height: 7px;
  }
}

.spin-lg .spin-dot{
  width: 45px;
  height: 45px;

  .spin-dot-item {
    width: 14px;
    height: 14px;
  }
}