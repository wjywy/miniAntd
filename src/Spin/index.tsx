import Button from './spin';

import '../style';
export default Button;
