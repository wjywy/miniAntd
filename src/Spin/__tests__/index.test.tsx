import { render } from '@testing-library/react';
import React from 'react';

import Spin from '../';

describe('Spin', () => {
  it('should render correctly', () => {
    const { getByTestId } = render(
        <Spin data-testid="Spin"></Spin>
    );
    expect(getByTestId('Spin')).toMatchSnapshot();
  });
});
