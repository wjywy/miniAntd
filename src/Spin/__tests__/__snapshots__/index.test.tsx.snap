// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spin should render correctly 1`] = `
<div
  aria-busy="true"
  aria-live="polite"
  class="spin spin-spinning"
  data-testid="Spin"
>
  <span
    class="spin-dot spin-dot-spin"
  >
    <i
      class="spin-dot-item"
    />
    <i
      class="spin-dot-item"
    />
    <i
      class="spin-dot-item"
    />
    <i
      class="spin-dot-item"
    />
  </span>
</div>
`;
